File: python/logbrew_py/src/logbrew_sdk/_timeline.py
"""Agent-readable timeline helpers for product and network milestones."""

from __future__ import annotations

import math
import re
from collections.abc import Mapping
from typing import Any
from urllib.parse import urlsplit

from logbrew_sdk import (
    ACTION_STATUSES,
    ActionAttributes,
    Metadata,
    MetadataValue,
    SdkError,
    compact_metadata,
    require_allowed_value,
    require_non_empty,
)

HTTP_METHOD_PATTERN = re.compile(r"^[A-Z][A-Z0-9_-]*$")


def create_product_action_attributes(
    action: str | Mapping[str, Any],
    *,
    metadata: Mapping[str, Any] | None = None,
) -> ActionAttributes:
    """Build privacy-safe action attributes for app-owned product milestones."""

    details = product_action_details(action)
    timeline_metadata = {
        "routeTemplate": sanitize_route_template(details.get("routeTemplate")),
        "sessionId": string_or_none(details.get("sessionId")),
        "traceId": string_or_none(details.get("traceId")),
        "screen": string_or_none(details.get("screen")),
        "funnel": string_or_none(details.get("funnel")),
        "step": string_or_none(details.get("step")),
    }
    return {
        "name": details["name"],
        "status": details["status"],
        "metadata": compact_metadata(
            {
                "source": "product.action",
                **dict(compact_metadata(metadata) or {}),
                **dict(compact_metadata(details.get("metadata")) or {}),
                **drop_none_values(timeline_metadata),
            }
        )
        or {},
    }


def create_network_milestone_attributes(
    request: str | Mapping[str, Any],
    *,
    metadata: Mapping[str, Any] | None = None,
) -> ActionAttributes:
    """Build privacy-safe action attributes for app-owned network milestones."""

    details = network_milestone_details(request)
    timeline_metadata = {
        "routeTemplate": details["routeTemplate"],
        "method": details["method"],
        "statusCode": details.get("statusCode"),
        "durationMs": details.get("durationMs"),
        "sessionId": string_or_none(details.get("sessionId")),
        "traceId": string_or_none(details.get("traceId")),
    }
    return {
        "name": details["name"],
        "status": details["status"],
        "metadata": compact_metadata(
            {
                "source": "network.milestone",
                **dict(compact_metadata(metadata) or {}),
                **dict(compact_metadata(details.get("metadata")) or {}),
                **drop_none_values(timeline_metadata),
            }
        )
        or {},
    }


def product_action_details(action: str | Mapping[str, Any]) -> dict[str, Any]:
    if isinstance(action, str):
        require_non_empty("product action name", action)
        return {"name": action, "status": "success"}
    if not isinstance(action, Mapping):
        raise SdkError("validation_error", "product action must be a string or object")

    require_non_empty("product action name", action.get("name"))
    status = action.get("status", "success")
    require_allowed_value("product action status", status, ACTION_STATUSES)
    return {
        "funnel": action.get("funnel"),
        "metadata": action.get("metadata"),
        "name": action["name"],
        "routeTemplate": action.get("routeTemplate"),
        "screen": action.get("screen"),
        "sessionId": action.get("sessionId"),
        "status": status,
        "step": action.get("step"),
        "traceId": action.get("traceId"),
    }


def network_milestone_details(request: str | Mapping[str, Any]) -> dict[str, Any]:
    if isinstance(request, str):
        return network_milestone_details({"routeTemplate": request})
    if not isinstance(request, Mapping):
        raise SdkError("validation_error", "network milestone must be a string or object")

    route_template = sanitize_route_template(request.get("routeTemplate"))
    require_non_empty("network milestone routeTemplate", route_template)
    method = normalize_http_method(request.get("method"))
    status_code = status_code_or_none(request.get("statusCode"))
    status = request.get("status", status_from_status_code(status_code))
    require_allowed_value("network milestone status", status, ACTION_STATUSES)
    duration_ms = non_negative_number_or_none("network milestone durationMs", request.get("durationMs"))
    name = request.get("name")
    if not isinstance(name, str) or not name.strip():
        name = f"network.{method.lower()} {route_template}"

    return {
        "durationMs": duration_ms,
        "metadata": request.get("metadata"),
        "method": method,
        "name": name,
        "routeTemplate": route_template,
        "sessionId": request.get("sessionId"),
        "status": status,
        "statusCode": status_code,
        "traceId": request.get("traceId"),
    }


def sanitize_route_template(route_template: Any) -> str | None:
    if route_template is None:
        return None
    if not isinstance(route_template, str):
        raise SdkError("validation_error", "routeTemplate must be a string")
    trimmed = route_template.strip()
    if not trimmed:
        return ""
    parsed = urlsplit(trimmed)
    if parsed.scheme or parsed.netloc:
        return parsed.path or "/"
    return re.split(r"[?#]", trimmed, maxsplit=1)[0] or "/"


def normalize_http_method(method: Any) -> str:
    value = "GET" if method is None else method
    if not isinstance(value, str) or not value.strip():
        raise SdkError("validation_error", "network milestone method must be a non-empty string")
    normalized = value.strip().upper()
    if HTTP_METHOD_PATTERN.fullmatch(normalized) is None:
        raise SdkError("validation_error", "network milestone method must be a valid HTTP method")
    return normalized


def status_code_or_none(value: Any) -> int | None:
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, int) or value < 100 or value > 599:
        raise SdkError("validation_error", "network milestone statusCode must be an integer from 100 to 599")
    return value


def status_from_status_code(status_code: int | None) -> str:
    if status_code is not None and status_code >= 400:
        return "failure"
    return "success"


def non_negative_number_or_none(label: str, value: Any) -> float | None:
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) or value < 0:
        raise SdkError("validation_error", f"{label} must be a non-negative number")
    return float(value) if isinstance(value, int) else value


def string_or_none(value: Any) -> str | None:
    return value if isinstance(value, str) and value.strip() else None


def drop_none_values(metadata: Mapping[str, MetadataValue | None]) -> Metadata:
    return {key: value for key, value in metadata.items() if value is not None}

File: go/logbrew/timeline.go
package logbrew

import (
	"fmt"
	"math"
	"net/url"
	"strings"
)

// ProductActionInput describes an app-owned product step that should be
// captured as an agent-readable action event.
type ProductActionInput struct {
	Name          string
	Status        string
	RouteTemplate string
	SessionID     string
	TraceID       string
	Screen        string
	Funnel        string
	Step          string
	Metadata      map[string]any
}

// NetworkMilestoneInput describes an app-owned API milestone that should be
// captured as an agent-readable action event.
type NetworkMilestoneInput struct {
	Name          string
	RouteTemplate string
	Method        string
	Status        string
	StatusCode    *int
	DurationMs    *float64
	SessionID     string
	TraceID       string
	Metadata      map[string]any
}

// CreateProductActionAttributes builds privacy-safe action attributes for a
// product milestone without automatic click capture or global app mutation.
func CreateProductActionAttributes(input ProductActionInput) (ActionAttributes, error) {
	if err := requireNonEmpty("product action name", input.Name); err != nil {
		return ActionAttributes{}, err
	}
	status := input.Status
	if status == "" {
		status = "success"
	}
	if err := requireAllowedValue("product action status", status, actionStatus); err != nil {
		return ActionAttributes{}, err
	}
	resultMetadata := timelineMetadata("product.action", input.Metadata, map[string]any{
		"routeTemplate": sanitizeRouteTemplate(input.RouteTemplate),
		"sessionId":     stringOrNil(input.SessionID),
		"traceId":       stringOrNil(input.TraceID),
		"screen":        stringOrNil(input.Screen),
		"funnel":        stringOrNil(input.Funnel),
		"step":          stringOrNil(input.Step),
	})
	return ActionAttributes{Name: input.Name, Status: status, Metadata: resultMetadata}, nil
}

// CreateNetworkMilestoneAttributes builds privacy-safe action attributes for
// an API milestone without patching HTTP clients or capturing payloads/headers.
func CreateNetworkMilestoneAttributes(input NetworkMilestoneInput) (ActionAttributes, error) {
	routeTemplate, err := requireRouteTemplate(input.RouteTemplate)
	if err != nil {
		return ActionAttributes{}, err
	}
	method, err := normalizeHTTPMethod(input.Method)
	if err != nil {
		return ActionAttributes{}, err
	}
	statusCode, hasStatusCode, err := statusCodeValue(input.StatusCode)
	if err != nil {
		return ActionAttributes{}, err
	}
	status := input.Status
	if status == "" {
		status = statusFromStatusCode(statusCode, hasStatusCode)
	}
	if err := requireAllowedValue("network milestone status", status, actionStatus); err != nil {
		return ActionAttributes{}, err
	}
	durationMs, hasDurationMs, err := nonNegativeNumberValue("network milestone durationMs", input.DurationMs)
	if err != nil {
		return ActionAttributes{}, err
	}
	name := strings.TrimSpace(input.Name)
	if name == "" {
		name = fmt.Sprintf("network.%s %s", strings.ToLower(method), routeTemplate)
	}
	timeline := map[string]any{
		"routeTemplate": routeTemplate,
		"method":        method,
		"sessionId":     stringOrNil(input.SessionID),
		"traceId":       stringOrNil(input.TraceID),
	}
	if hasStatusCode {
		timeline["statusCode"] = statusCode
	}
	if hasDurationMs {
		timeline["durationMs"] = durationMs
	}
	resultMetadata := timelineMetadata("network.milestone", input.Metadata, timeline)
	return ActionAttributes{Name: name, Status: status, Metadata: resultMetadata}, nil
}

func timelineMetadata(source string, local map[string]any, timeline map[string]any) map[string]any {
	result := map[string]any{"source": source}
	for key, value := range compactMetadata(local) {
		result[key] = value
	}
	for key, value := range timeline {
		if value != nil {
			result[key] = value
		}
	}
	return compactMetadata(result)
}

func requireRouteTemplate(routeTemplate string) (string, error) {
	sanitized := sanitizeRouteTemplate(routeTemplate)
	if err := requireNonEmpty("network milestone routeTemplate", sanitized); err != nil {
		return "", err
	}
	return sanitized, nil
}

func sanitizeRouteTemplate(routeTemplate string) string {
	trimmed := strings.TrimSpace(routeTemplate)
	if trimmed == "" {
		return ""
	}
	parsed, err := url.Parse(trimmed)
	if err == nil && (parsed.IsAbs() || parsed.Host != "") {
		if parsed.Path == "" {
			return "/"
		}
		return parsed.Path
	}
	if index := strings.IndexAny(trimmed, "?#"); index >= 0 {
		trimmed = trimmed[:index]
	}
	if trimmed == "" {
		return "/"
	}
	return trimmed
}

func normalizeHTTPMethod(method string) (string, error) {
	value := strings.TrimSpace(method)
	if value == "" {
		value = "GET"
	}
	normalized := strings.ToUpper(value)
	if !isValidHTTPMethod(normalized) {
		return "", &SdkError{Code: "validation_error", Message: "network milestone method must be a valid HTTP method"}
	}
	return normalized, nil
}

func isValidHTTPMethod(method string) bool {
	if method == "" {
		return false
	}
	for index, char := range method {
		if index == 0 && (char < 'A' || char > 'Z') {
			return false
		}
		if !((char >= 'A' && char <= 'Z') || (char >= '0' && char <= '9') || char == '_' || char == '-') {
			return false
		}
	}
	return true
}

func statusCodeValue(statusCode *int) (int, bool, error) {
	if statusCode == nil {
		return 0, false, nil
	}
	if *statusCode < 100 || *statusCode > 599 {
		return 0, false, &SdkError{Code: "validation_error", Message: "network milestone statusCode must be an integer from 100 to 599"}
	}
	return *statusCode, true, nil
}

func statusFromStatusCode(statusCode int, present bool) string {
	if present && statusCode >= 400 {
		return "failure"
	}
	return "success"
}

func nonNegativeNumberValue(label string, value *float64) (float64, bool, error) {
	if value == nil {
		return 0, false, nil
	}
	if math.IsNaN(*value) || math.IsInf(*value, 0) || *value < 0 {
		return 0, false, &SdkError{Code: "validation_error", Message: fmt.Sprintf("%s must be a non-negative number", label)}
	}
	return *value, true, nil
}

func stringOrNil(value string) any {
	if strings.TrimSpace(value) == "" {
		return nil
	}
	return value
}
